rust/agama-lib/src/utils/transfer/file_finder.rs

```diff
@@ -0,0 +1,121 @@
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE LLC about this file by physical or electronic mail, you may
+// find current contact information at www.suse.com.
+
+use std::{
+    io::Write,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use super::{
+    file_systems::{FileSystem, FileSystemsList},
+    TransferError, TransferResult,
+};
+
+/// Finds a file in a set of file systems and copies its content.
+#[derive(Default)]
+pub struct FileFinder {}
+
+impl FileFinder {
+    /// Searchs for a file in the given file systems and copies its content to the given writer.
+    ///
+    /// * `file_systems`: file systems to search in.
+    /// * `file_name`: file name.
+    /// * `writer`: where to write the contents.
+    pub fn copy_from_file_systems(
+        &self,
+        file_systems: &FileSystemsList,
+        file_name: &str,
+        writer: &mut impl Write,
+    ) -> TransferResult<()> {
+        for fs in file_systems.to_vec().iter() {
+            if self.copy_from_file_system(fs, &file_name, writer).is_ok() {
+                return Ok(());
+            }
+        }
+        Err(TransferError::FileNotFound(file_name.to_string()))
+    }
+
+    /// Copies the file from the file system to the given writer.
+    ///
+    /// * `file_systems`: file systems to search in.
+    /// * `file_name`: file name.
+    /// * `writer`: where to write the contents.
+    pub fn copy_from_file_system(
+        &self,
+        file_system: &FileSystem,
+        file_name: &str,
+        writer: &mut impl Write,
+    ) -> TransferResult<()> {
+        const DEFAULT_MOUNT_PATH: &str = "/run/agama/mount";
+        let default_mount_point = PathBuf::from(DEFAULT_MOUNT_PATH);
+        let mount_point = file_system
+            .mount_point
+            .clone()
+            .unwrap_or(default_mount_point);
+
+        println!("Searching {} in {}", &file_name, &file_system.block_device);
+
+        if !file_system.is_mounted() {
+            Self::mount_file_system(&file_system, &mount_point)?;
+        }
+
+        let file_name = file_name.strip_prefix("/").unwrap_or(file_name);
+        let source = mount_point.join(&file_name);
+        let result = Self::copy_file(source, writer);
+
+        if !file_system.is_mounted() {
+            Self::umount_file_system(&mount_point)?;
+        }
+
+        result
+    }
+
+    /// Mounts the file system on a given point.
+    ///
+    /// TODO: move this logic to the FileSystem struct.
+    fn mount_file_system(file_system: &FileSystem, mount_point: &PathBuf) -> TransferResult<()> {
+        std::fs::create_dir_all(mount_point)?;
+        let output = Command::new("mount")
+            .args([file_system.device(), mount_point.display().to_string()])
```

Comment:
what about using by default `ro` to avoid accidental overwrite? and maybe for some devices you will need `-o loop`